add RFC for debuginfo. (#107)
prjunnamed/prjunnamed: rfcs/0000-debuginfo.md

   - if a net is preserved, it must be possible to compute its value from the current state of the design
   - if a register or memory is preserved, it must be possible to modify its current value, and have the design use the new value for all purposes
     - corollary: even if a preserved register is obviously always 0 (it is initialized to 0, and the D input is tied to 0), the synthesis flow must not delete it, and must not assume its value to be 0 in any optimization passes (since the debugger may want to set it to 1)

3. Preserve all: all named nets and registers are considered to be preserved as above.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

The debug information model in prjunnamed is composed of three interconnected layers:

1. A tree of *scopes* describing the hierarchy of the design (corresponding to HDL modules and blocks).  Scopes provide the skeleton to which named objects are attached.
2. Named objects:

   - *named nets*, which allow a value to be observed
   - *named signals* and *named registers*, which allow a value to be read or written

>    - *named cells*, which allow specific target cell instances to be accessed (in unspecified ways)

Reviewer: Soo... signals = can be read, registers = can be written? If so, what's the difference between nets and signals?Or do you mean something else?

Also, what is the distinction you're making between "observed" and "read" here?

---

Okay, having read further, I'd suggest the following wording:

```suggestion
   - *named nets*, which allow a value to be observed
   - *named signals* and *named registers*, which are points that allow a value to be read and/or written
   - *named cells*, which allow specific target cell instances to be located for the purpose of access by other tooling (the nature of which we leave unspecified)
```

Still not sure about the implications of the "read" vs "observed" wording though.

Author: yeah I'll just replace "observe" with "read" everywhere; also, I'll add a section about a "debug interface" for the simulator could look like

3. *Data types*, which describe how the values of named nets and registers are to be interpretted or displayed.

## Scopes

> The prjunammed IR is flat by nature: there are no modules, only a single design.  Scopes are a type of metadata that describes the hierarchy of the design as originally elaborated, and can be used to associate named objects and IR cells with their location in it.  Every scope and named object has a *hierarchical name*, which is unique in the design.  The hierarchical name is essentially a path from the root scope, consisting of names and indices of all parent scopes.

Reviewer: > with their location in it.

This `it` doesn't properly dereference for me, linguistically. "their location in the user's source code" maybe?

Author: "the hierarchy" actually; fixed

Scopes come in three kinds:

- the singular implicit *root scope*, which has an empty hierarchical name (corresponds to `$root` in verilog)
- a *named scope*, which is contained in another scope, and is identified by a unique *name* within it.  A name can be any UTF-8 string, though it is usually expected to be a valid identifier.

> - an *indexed scope*, which is contained in another scope, and is identified by a unique *index* within it.  The index can be an arbitrary `i64` integer.

Reviewer: After introducing "kinds", I'd immediately at least mention "types", otherwise the example is confusing, and by the time you get to your definition of "types" down below, the previous discussion of "kinds" got upcasted to the `{type, kind, gender}` equivalence class

Reviewer 2: I propose eliminating scope kinds and instead saying that a scope *name* can be either an *identified name* or an *indexed name*. I think this is better regardless of introducing the term "kinds" because the form of the name doesn't really influence anything besides how things are displayed/serialized; they do not need to be treated different by tooling, so there isn't a good reason to use kinds.

(The concept of a root scope seems pedagogic and equivalent to allowing top-level scope with no parent, so that shouldn't be an issue either.)

!5 = scope "inst1" in=!3 type=module("abc")
!6 = scope #1 in=!2 type=block
!7 = scope "inst0" in=!6 type=module("abc")
!8 = scope "inst1" in=!6 type=module("abc")
!9 = scope #2 in=!2 type=block
!10 = scope "inst0" in=!9 type=module("abc")
!11 = scope "inst1" in=!9 type=module("abc")
# a named scope representing a module instance array
!12 = scope "inst2" in=!1 type=array
# an indexed scopr representing a specific instance in the array
!13 = scope #0 in=!12 type=module("xyz") 
!14 = scope "a" in=!13 type=block
!15 = scope #1 in=!12 type=module("xyz") 
!16 = scope "a" in=!15 type=block
!17 = scope #2 in=!12 type=module("xyz") 
!18 = scope "a" in=!17 type=block
```

The hierarchical names of the scopes (written in Verilog convention) would be:

- `!1`: top
- `!2`: top.loop
- `!3`: top.loop[0]
- `!4`: top.loop[0].inst0
- `!5`: top.loop[0].inst1
- `!6`: top.loop[1]
- `!7`: top.loop[1].inst0
- `!8`: top.loop[1].inst1
- `!9`: top.loop[2]
- `!10`: top.loop[2].inst0
- `!11`: top.loop[2].inst1
- `!12`: top.inst2
- `!13`: top.inst2[0]
- `!14`: top.inst2[0].a
- `!15`: top.inst2[0]
- `!16`: top.inst2[0].a
- `!17`: top.inst2[0]
- `!18`: top.inst2[0].a

> Scopes can have one of several predefined types (such as module or block), which are advisory and carry no semantics to prjunnamed itself.  They can also have source locations.  Scopes are metadata and as such can be attached to cells to signify their origin, but their main role is to contain named objects.

Reviewer: ~~I'm assume you're intentionally not specifying what the types should be here beyond what occurs in the example?~~

Do mention that the full list of available types is left for the reference-level guide section, otherwise it feels almost underspecified as compared to what came before.

Reviewer: > [scope types] are advisory and carry no semantics to prjunnamed itself.

This is in conflict with the reference-level explanation specifying semantic constraints on `struct` and `interface` scopes.

Author: .... yeah, removed it

```
!2 = net "y" width=4 in=!1 keep
%3:0 = name %2:4 !2
```

The name is associated with a value, not with the cell driving it.  Even if marked as "keep", the flow is free to optimize the logic driving it in any way that preserves its semantics.  For example, if it determines that a net always have a constant value, one may end up with a named constant:

```
!2 = net "y" width=4 in=!1
%3:0 = name 0101 !2
```

### Named signals

A *named signal* is a variant of a named net that can also be used to *drive* a value into the design.  In normal circumstances, it acts as a buffer forwarding its input to output.  However, it can be overriden by an external entity to drive a different value:

```
wire[3:0] x = a + b;
wire[3:0] y = x + c;
```

```
%0:4 = [...] # a
%1:4 = [...] # b
%2:4 = [...] # c
# add a + b
%3:5 = adc %0:4 %1:4 0
# the named signal, consuming the above value
%4:4 = signal %3:4 "x" in=!1
# add x + c
%5:5 = adc %4:4 %2:4 0
```

> A named signal acts as an optimization barrier: the flow cannot assume that the value of `x` flowing into the second addition is the same as the result of the first addition, as it could be overriden by something external.

Reviewer: How does acting as an optimization barrier relate to the fact that signals are the recommended way of representing ports between modules? Perhaps the behavior of the optimization barrier should only happen if the signal is marked with the `keep` attribute??

Author: Ports can be represented in three different ways depending on your debugging needs:

- nets (read-only)
- signals (read or write)
- registers (read or write)

The usual (non-debug) way to use them would be:

- internal ports: represented as nets (or, opportunistically, as registers if there happens to be one driving an output port)
- top-level outputs: represented as `signal` (with its value unused)
- top-level inputs: represented as `signal` (with input of `X`, as it's expected to be overriden)

The full-debug way would represent everything as signals.

Author: added the explanation above into the RFC text

> Named signals are a debugging tool that is not normally used, and should only be emitted by explicit user request.  They always act as a strong reference, and will not be optimized out.

Reviewer: I'd include a sentence or two about how a named signal could be synthesized (marked LUTs whose contents can be modified at runtime by the debugger? cute little mux + flop combo that stays transparent when not in use?)

While it is not this RFC's place to go into the details, something like this could help the reader conceptualize this feature.

Author: I think this is an unresolved question.

the proper lowering of a signal depends on what you want to do with the design:

- do post-synth simulation or verification: just leave the signal as-is
- do in-circuit debugging: indeed, one of the ways you mentioned above (which one depends on target)

this puts it in the same category as the "do we need this register to be writable through ICD, or have its initial value changed through bitstream manipulation" unresolved question we currently have

Reviewer 2: I am not sure that named signals justify the complexity they add.
- For primary inputs/outputs, we already have the appropriate cells with much narrower semantics and no impact on optimization, so it doesn't seem like that *needs* to be interpreted as named signals.
- Without named signals it would not be possible to e.g. drive a combinational signal within a design like Verilog's `force` construct allows it. But I'm not sure this is particularly needed given prjunnamed's scope: Amaranth explicitly disallows it and I've not heard anybody complain.
- Would a barrier cell combined with a normal `name keep` cell not provide the same functionality? An explicit barrier cell would be valuable on its own, and if this functionality is only provided via `signal` then I expect frontends would have to come up with a `signal` just to create a barrier, which seems bad.
- How does a signal differ from an always-enabled latch semantically? (I'm assuming here that a latch can have an associated `register` cell which allows one to override its value.)

> ### Ports

Reviewer: Ports deserve to be mentioned in the list above, otherwise there's the moment of "wait, weren't we supposed to be doing registers?"

Author: added (though kinda in a clumsy way I think)

> Ports between modules in the hierarchy are represented by named nets or signals with specifiic annotations:

Reviewer: This kinda feels like "wait, where did ports come from, they weren't mentioned before". Maybe
```suggestion
### Representing ports

Ports between modules in the hierarchy are represented by named nets or signals with specific annotations:
```

```
module top;
    wire ia, iy;
    inv inst0(.a(ia), .y(iy));
endmodule

module inv(input a, output y);
    assign y = !a;
endmodule
```

```
!1 = scope "top" type=module("top")
!2 = scope "inst0" type="module("inv") in=!1
%0 = [...] # ia
%1 = signal %0 "a" in=!2 input
%2 = not %1
%3 = signal %2 "y" in=!2 output
```

Likewise, named signals are used for top-level ports, obsoleting the current `input` and `output` cells:

```
module top(input a, output y);
    assign y = !a;
endmodule
```

```
!1 = scope "top"
# top-level input port represented as a signal cell with undefined input

> %0 = signal X "a" in=!1 input

Reviewer: Suspicious.

Author: ... meow?

Reviewer: Just that the use of `X` felt kinda different from the usual semantics.

Author: I mean, it's the canonical "don't care" value to be connected somewhere that should be unused

%1:1 = not %0
# top-level output port represented as a signal cell with unused output

> %2:1 = signal %1 "y" in=!2 output

Reviewer: As a sidenote, the `in=` parameter feels kind of ambiguous. Perhaps `within=` would be less likely to be confused with "in as in input"?

Alternatively, we could have `scope=` for things that aren't themselves scopes, and `parent=` for non-top-level scopes. But that's not very orthogonal...

Author: true, it is confusing

I don't particularly like `within=` (for some reason I think of it as "the value is within this range of values" annotation), so I'll go with `parent=` and `scope=`

Reviewer: Hm, what about `inside=`?

Like named nets, by default named registers are considered to be "weak references" and can be destroyed by optimizations whenever convenient.  This can be changed with one of the "keep" annotations:

1. `keep`: the register must remain both readable and writable by external entities.  Every bit of the register must be mapped to at least one bit of storage in the netlist, and (like `signal`) the output of the register must be treated as an optimization barrier (to ensure an externally written value is properly observed)
2. `keep_write`: the register must remain writable by external entities, but read access is not required.  Behaves like `keep`, but allows the synthesis tool to delete bits that are unused by the netlist (an external entity can write to them, which is treated as a no-op).
3. `keep_read`: the register must remain readable by external entities, but write access is not required.  If convenient for optimization, this register (or any subset of its bits) can be "decayed" into a named net with the `keep` annotation.

> 4. Default: the register or any subset of its bits can be deleted, or decayed into a named net.

Reviewer: ```suggestion
4. Default: the register or any subset of its bits can be deleted, or decayed into a named net, as necessary.
```

}
!2 = reg "pid" width=4 type=!1
```

### Structure types

Display the value by splitting it into fields, then recursing into fields.  Has an associated width, and can only be used on values of that width.  The starting offset of each field is given explicitly.  Unions can be described as structs where each field happens to be at offset 0.

```
!1 = type unsigned
!2 = type struct #16 {
    field "b" start=#0 width=#5 type=!1
    field "g" start=#5 width=#6 type=!1
    field "r" start=#11 width=#5 type=!1
}
```

### Array types

Used for memories, as well as packed arrays within structs.

```
reg [31:0] gprs[32];
```

```
!1 = type bitvec
!2 = type array width=#32 depth=#32 type=!1
!3 = reg "gprs" width=#1024 type=!2
%0:_ = memory width=#32 depth=#32 {
    name !3:1024
    [...]
}
```

### Tagged union

Compares the value against a list of bit patterns, takes the first matching one, then decodes the value as a structure according to the matched variant.

```
!1 = type union #16 {
    011000xxxxxxxxxx = "ADD" {
        field "dst" start=#0 width=#5
        field "src" start=#5 width=#10
    }
    1011xxxxxxxxxxxx = "JUMP" {
        field "target" start=#0 width=#12

>     }n

Reviewer: ```suggestion
    }
```

Cat on keyboard incident?

Author: meow.

>     [..]

Reviewer: ```suggestion
    [...]
```

}
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

The following metadata items are added to the IR:

- `scope <name or index> [in=<scope>] [type=<scope type>] [src=<source>]` (slight modification to existing metadata)

>   - `type=module[(<name>)]`
>   - `type=block`

Reviewer: ```suggestion
  - `type=module[(<name>)]`
  - `type=block`
  - `type=struct`
```

- `input` (replaced by `signal`)
- `output` (replaced by `signal`)
- `name` (replaced by the new `name` cell with different semantics)
- `debug` (replaced by `name`)

Furthermore, an optional `reg` metadata reference is added to the following cells:

- `dff`
- `memory` (memory contents and synchronous read registers)
- `target`

## Scopes

Scopes must form a tree.  A given hierarchical name must correspond to at most one scope or named object.

The IR puts no constraints on the names beyond being valid UTF-8 strings.

Usually there will be exactly one top-level module scope containing all other scopes and named objects in the design (the `top` module).  However, this is not a hard requirement.  It is allowed to have multiple top-level scopes. It is also allowed, though heavily discouraged, to put named objects directly inside the root scope.

The following scope types are defined:

- `module` or `module(<name>)`: represents a Verilog `module` instance, a VHDL `entity` instance, or a similar construct; optionally, the name of the module type may be given for informational purposes
- `block`: represents a block within a module; corresponds to a Verilog `block`

> - `struct`: represents a struct-typed signal, wire, or similar construct that got split into individual named nets or signals by the backend

Reviewer: No optional name for structs?

Author: .... I'll add one

I think I may have been slightly biased by the fact that representing SystemVerilog type names would be incredibly complex (there may be generics involved)

- `interface` or `interface(<name>)`: represents a grouping of module's input / output ports; like `module`, the name of the interface type may be given for informational purposes
- `array`: represents an array of blocks, interfaces, or module instances

It is expected that indexed scopes will only be contained within a scope of type `array`, which may be a named scope or another (nested) indexed scope.  Likewise, it is expected that `array` scopes will contain only indexed scopes and named objects.  However, this is not a hard requirement.

The IR has two distinct tools to represent composite data types: the `struct` scope type, which represents fields as individual IR objects, and the `type struct`, which represents the entire structure as a single entity, treating fields as a matter of display processing.  The HDL frontend makes the choice between them as appropriate (the two choices roughly correspond to `packed` and `unpacked` SystemVerilog types).

The `struct` scope type has special constraints:

- it can only contain (directly or indirectly): named nets, named signals, named registers, `struct` scopes, `array` scopes

The `interface` type has special constraints:

> - it can only be contained within a `module` or another `interface`, directly or inside `array` scopes (cannot be contained within a `block`)

Reviewer: ```suggestion
- it can only be contained within a `module` or another `interface`, directly or inside `array` scopes (cannot be contained within a `block` or `struct`)
```

> - it can only contain (directly or indirectly): named nets, named signals, `interface` scopes, `array` scopes, `struct` scopes (cannot contain `block` or `module scopes, nor other kinds of cells)

Reviewer: ```suggestion
- it can only contain (directly or indirectly): named nets, named signals, `interface` scopes, `array` scopes, `struct` scopes (cannot contain `block` or `module` scopes, nor other kinds of cells)
```

> - it can only contain (directly or indirectly): named nets, named signals, `interface` scopes, `array` scopes, `struct` scopes (cannot contain `block` or `module scopes, nor other kinds of cells)

Reviewer: To what extent do these constraints give us concrete advantages, and to what extent is this "your language must be sufficiently similar to Verilog if you want debuginfo"

Author: these constraints aren't really designed for Verilog in the first place! (SV interfaces would be represented as `module`, not `interface`, as their design is uhhh kinda batshit; though I guess with enough post-processing we could reconstruct some reasonable-looking ports out of them most of the time)

my reasoning was:

- we want to have netlist exporters that will represent the netlist as, say, hierarchical Verilog whether it wants to or not
- ie. we want to reconstruct a proper hierarchy with proper ports between modules
- for that, we want to know what the cross-module ports are (use the existing ones if available, add new ports as necessary to route nets between modules)
- for that, we need to associate a port with its module
  - the obvious way would be to consider the immediately containing scope to be a port's module
  - however, this breaks down for ports that are unpacked structs, or interfaces
  - therefore, we should explicitly mark scopes that are unpacked structs or interfaces (and so `scope type=` is born, and consider them "transparent" for the purpose of finding enclosing module
  - so the lookup halts on a `module` scope, goes up on `interface`, `struct`, and `array`, okay
  - ... but what if a port is inside a `block` scope? blocks should not *have* ports, there's nowhere to put them
  - let's just disallow this case
    - ports cannot be directly or indirectly in `block` without intervening `module`
    - let's also disallow having `interface` inside a `block` at all, as it's only meant to contain ports
  - while we're at it, let's also add constraints on `struct` that reflect its "Plain Old Data with no surprises" intent

The whole thing also ties into the future "IR linker" plans: we want to be able to take a `Design` and plop it into another design in place of an `Instance` cell. for that, we need top-level ports to have a well-defined structure that can then be cross-referenced with the port list on the `Instance` (the current definition of `Instance` doesn't allow hierarchical names on the ports that would be required to support `array`/`interface`/`struct` ports, but maybe it should).  Further, if the "cross-language elaboration loop" future plans ever pan out, we'll want a well-defined model of "module interfaces", so that the top-level elaborator can extract the interface of a just-elaborated submodule and send it to the language frontend so it can do proper type coercion.

ie. this is implicit, but I want us to be able to extract an "interface type" from a design, which would go like this:

```
type DesignType = Interface;
struct Interface {
    fields: Map<String, InterfaceType>,
}
enum InterfaceType {
    Input(DataType),
    Output(DataType),
    Interface(Interface),
    Array(InterfaceType, usize),
}
struct Struct {
    fields: Map<String, DataType>,
}
enum DataType {
    Struct(Struct),
    Array(DataType, usize),
    Bitvec(usize),
}
```

... which I now realize needs more constraints. ugh.

> Netlist exporters are expected to deal with arbitrary names and scope hierarchies.  However, scope structures and names not representable in the target format or language may not be preserved (scopes may be flattened or inserted as necessary to legalize the hierarchy; names may be mangled, escaped, or stripped)  Therefore, unusual hierarchy structures and identifier characters should be avoided.

Reviewer: ```suggestion
Netlist exporters are expected to deal with arbitrary names and scope hierarchies.  However, scope structures and names not representable in the target format or language may not be preserved (scopes may be flattened or inserted as necessary to legalize the hierarchy; names may be mangled, escaped, or stripped). Therefore, unusual hierarchy structures and identifier characters should be avoided.
```

# Unresolved questions
[unresolved-questions]: #unresolved-questions

How should the metadata be stored in memory?  It is clear we want to be able to look up objects by their names, and find all `name` / `dff` / ... cells referencing a given named net or named register.  This will require maintaining an index.

What should be the exact semantics of named cells, if any?  Is it allowed and/or useful to duplicate a named cell?

What should be the exact syntax for the new metadata?  In particular, storage elements will require referencing (a potentially very large amount of) permuted named register bits.

Should we have a way to specify `keep` attributes that only apply to some part of the flow?  (example usecase: keep wires represented as named signals until the constraint file is parsed, convert them to named nets afterwards).

This RFC subsumes the current `input` and `output` cells by the unified named object model.  However, I/O values (used for bidirectional toplevel ports) are left untouched, despite also being identified by name.  Is there a good way to include them in the model?  Can they be made less special in some way?

> The `signal` cell provides an optimization barrier, which is a primitive that currently cannot be obtained otherwise.  Likewise, a named net with `keep` can prevent an arbitrary logic cone from being deleted.  Should we have cells that provide these semantics without requiring a name?

Reviewer: Well, that's kind of a theme for the project, isn't it.

   - if a net is preserved, it must be possible to compute its value from the current state of the design
   - if a register or memory is preserved, it must be possible to modify its current value, and have the design use the new value for all purposes
     - corollary: even if a preserved register is obviously always 0 (it is initialized to 0, and the D input is tied to 0), the synthesis flow must not delete it, and must not assume its value to be 0 in any optimization passes (since the debugger may want to set it to 1)

3. Preserve all: all named nets and registers are considered to be preserved as above.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

The debug information model in prjunnamed is composed of three interconnected layers:

1. A tree of *scopes* describing the hierarchy of the design (corresponding to HDL modules and blocks).  Scopes provide the skeleton to which named objects are attached.
2. Named objects:

   - *named nets*, which are points that allow a value to be read

>    - *named signals* and *named registers*, which are points that allow a value to be read or written

Reviewer 2: What is a signal and how is it different from a register? This is explained very far down below but I think it should be introduced here for clarity.

   - *named cells*, which allow specific target cell instances to be located for the purpose of access by other tooling (the nature of which we leave unspecified)
   - *ports*, which are represented as named nets, signals, or registers with a special annotation

> 3. *Data types*, which describe how the values of named nets and registers are to be interpretted or displayed.

Reviewer 2: Since the types place no constraints on the values, perhaps it would be clearer to call these *data formats*?

Author: that is a good idea

## Scopes

> The prjunammed IR is flat by nature: there are no modules, only a single design.  Scopes are a type of metadata that describes the hierarchy of the design as originally elaborated, and can be used to associate named objects and IR cells with their location in the hierarchy.  Every scope and named object has a *hierarchical name*, which is unique in the design.  The hierarchical name is essentially a path from the root scope, consisting of names and indices of all parent scopes.

Reviewer 2: Would it not be a problem during frontend development that you aren't quite sure what type of object you need to create or where, and you end up with both e.g. a `reg` and a `signal` with the same name, resulting in an invalid netlist?

Reviewer 2: How will you handle a situation where someone does e.g.

```
a = Signal(10)
m.d.comb += a[0].eq(i)
m.d.sync += a[1:].eq(a)
```

since this seems like it would require the same name to be assigned both the `reg` and `net` flavors? Does the `lsb=` parameter solve this? It would be good to have an explicit example somewhere in the RFC.

Author: So this is something I'm not very happy with in the current RFC and I need to rethink that. 

As it stands, the RFC already allows you to define a `reg` and use (some bits of) it via a `name` cell instead of an actual register (and it is necessary for representing partially optimized out registers). In the above case, you could just have the frontend emit such a thing directly.

... which makes me think that maybe the `reg` and `net` objects are just redundant and should be merged together.

However, this doesn't work for `signal`s. So maybe we also need to make *those* work more like registers instead?

# a named scope representing a module instance array
!12 = scope "inst2" parent=!1 type=array
# an indexed scopr representing a specific instance in the array
!13 = scope #0 parent=!12 type=module("xyz") 
!14 = scope "a" parent=!13 type=block
!15 = scope #1 parent=!12 type=module("xyz") 
!16 = scope "a" parent=!15 type=block
!17 = scope #2 parent=!12 type=module("xyz") 
!18 = scope "a" parent=!17 type=block
```

The hierarchical names of the scopes (written in Verilog convention) would be:

- `!1`: top
- `!2`: top.loop
- `!3`: top.loop[0]
- `!4`: top.loop[0].inst0
- `!5`: top.loop[0].inst1
- `!6`: top.loop[1]
- `!7`: top.loop[1].inst0
- `!8`: top.loop[1].inst1
- `!9`: top.loop[2]
- `!10`: top.loop[2].inst0
- `!11`: top.loop[2].inst1
- `!12`: top.inst2
- `!13`: top.inst2[0]
- `!14`: top.inst2[0].a
- `!15`: top.inst2[0]
- `!16`: top.inst2[0].a
- `!17`: top.inst2[0]
- `!18`: top.inst2[0].a

Scopes can have one of several predefined types (module, block, interface, struct, array).  They can also have source locations.  Scopes are metadata and as such can be attached to cells to signify their origin, but their main role is to contain named objects.

> Module scopes are the main building blocks of hierarchy.  Block scopes represent subdivisions of logic within a module.  Interface scopes are used to group together related module ports.  Struct scopes are used to represent a group of nets, signals or registers that was a single named struct-typed object in the HDL.  Array scopes represent arrays of any of the other type of scopes.

Reviewer 2: This would be easier to read as a list.

Reviewer 2: Are there legality constraints on array scopes, e.g. can an array scope contain named scopes, or is it indexed only? Can struct scopes contain indexed scopes?

Author: I didn't see a hard reason to include such constraints, but yeah, it may be a good idea to have those (unless there's some widely used HDL that would actually want to violate these constraints?)

Reviewer 2: I strongly lean towards "add every constraint that seems like the obvious thing, relax later if we must" considering we aren't in any danger of having 1.0 any time soon (where the *emitted* format must be backward-compatible, rather than the *consumed* format accepted by the previous revisions needs to continue being consumable).

endmodule
```

```
!1 = scope "top"
%0:4 = [...] # a
%1:4 = [...] # b
%2:5 = adc %0:4 %1:4 0
# the metadata declares a net of width 4 in a scope
!2 = net "y" width=4 scope=!1
# the name cell is used to actually associate a value with the named net
%3:0 = name %2:4 !2
```

Named nets may be *unavailable*: if the value has been optimized out, it is no longer possible to read it from the design state.  This is represented by a net metadata without an associated name cell.

A name cell may be *partial*, and describe a slice of the associated named net.  This is used where the value may be only partially known:

```
# as above, but two top bits have been optimized out
%2:3 = adc %0:2 %1:2 0
!2 = net "y" width=4 scope=!1
%3:0 = name %2:2 !2+0:2
```

There may be several name cells associated with a net, to cover disjoint slices of a partially known value.

Normally, a name cell is effecively a "weak reference" to a value: the flow can delete a name, or any subset of its bits, if it is unused or stands in the way of optimization.  To prevent a value from being optimized out (making the name cell a strong reference), a net may be marked as "keep":

```
!2 = net "y" width=4 scope=!1 keep

> %3:0 = name %2:4 !2

Reviewer 2: Should this be a property of the *metadata* or the *cell*? It feels like debug metadata should not affect how the netlist is optimized, therefore `keep` would go on the cell.

Author: I thiiink having them on cells would be fine

```
!2 = net "y" width=4 scope=!1
%3:0 = name 0101 !2
```

### Named signals

A *named signal* is a variant of a named net that can also be used to *drive* a value into the design.  In normal circumstances, it acts as a buffer forwarding its input to output.  However, it can be overriden by an external entity to drive a different value:

```
wire[3:0] x = a + b;
wire[3:0] y = x + c;
```

```
%0:4 = [...] # a
%1:4 = [...] # b
%2:4 = [...] # c
# add a + b
%3:5 = adc %0:4 %1:4 0
# the named signal, consuming the above value
%4:4 = signal %3:4 "x" scope=!1
# add x + c
%5:5 = adc %4:4 %2:4 0
```

A named signal acts as an optimization barrier: the flow cannot assume that the value of `x` flowing into the second addition is the same as the result of the first addition, as it could be overriden by something external.

Named signals are a debugging tool that is not normally used, and should only be emitted by explicit user request.  They always act as a strong reference, and will not be optimized out.

> ```

Reviewer 2: This typo breaks the RFC text below it.

Like named nets, by default named registers are considered to be "weak references" and can be destroyed by optimizations whenever convenient.  This can be changed with one of the "keep" annotations:

1. `keep`: the register must remain both readable and writable by external entities.  Every bit of the register must be mapped to at least one bit of storage in the netlist, and (like `signal`) the output of the register must be treated as an optimization barrier (to ensure an externally written value is properly observed)
2. `keep_write`: the register must remain writable by external entities, but read access is not required.  Behaves like `keep`, but allows the synthesis tool to delete bits that are unused by the netlist (an external entity can write to them, which is treated as a no-op).
3. `keep_read`: the register must remain readable by external entities, but write access is not required.  If convenient for optimization, this register (or any subset of its bits) can be "decayed" into a named net with the `keep` annotation.

> 4. Default: the register or any subset of its bits can be deleted, or decayed into a named net, as necessary.

Reviewer 2: There is nothing wrong with this lattice conceptually, but I wonder how much of it will actually be used. Who will emit the `keep_read` and `keep_write` annotations? Humans are pretty unlikely to care a lot about fine grained semantics of something that is used pretty much entirely for debugging anyway; also this is a very prjunnamed-specific feature so user expectations will also play a factor in it likely not being used much.

I suspect that no-keep / keep is a sufficient taxonomy in practice. It is also backwards-compatible to introduce the other two options, so I propose simplifying the taxonomy unless there's a good reason to keep it.

Author: `keep_write` was motivated by the particular usecase of late-bound BRAM initialization: it is OK to remove unused bits of it, but it is not OK to rely on its initial contents. For this case, I'd assume we would have some high-level macro in HDL that defines a late-bound BRAM and applies the right attributes one way or another.

`keep_read` was added solely for symmetry, I don't have a good motivation for it.

Reviewer 2: I agree `keep_write` is valuable. Do you think we could split this 4-value enum into a pair of flags? Then I think the chance that some of the 4 valus will fly under the radar when writing a pass are somewhat lower, and also we don't have features added without a specific use case in mind.

Author: tbh I thought we'd just internally represent these four values as two bits, and show it like that for display purposes

Reviewer 2: Works; my worry here is just that some values would be under-tested.

### Representing ports

Ports between modules in the hierarchy are represented by named nets, signals, or registers with specific annotations:

```verilog
module top;
    wire ia, iy;
    inv inst0(.a(ia), .y(iy));
endmodule

module inv(input a, output y);
    assign y = !a;
endmodule
```

```
!1 = scope "top" type=module("top")
!2 = scope "inst0" type="module("inv") parent=!1
%0:1 = [...] # ia
%1:1 = signal %0 "a" scope=!2 input
%2:1 = not %1
%3:1 = signal %2 "y" scope=!2 output
```

Likewise, named signals are used for top-level ports, obsoleting the current `input` and `output` cells:

```verilog
module top(input a, output y);
    assign y = !a;
endmodule
```

```
!1 = scope "top"
# top-level input port represented as a signal cell with undefined input
%0:1 = signal X "a" scope=!1 input
%1:1 = not %0
# top-level output port represented as a signal cell with unused output
%2:1 = signal %1 "y" scope=!2 output
```

A named register can be used to represent ports when an output is directly driven by a register (such as Verilog `output reg ...`).

> The way to represent ports depends on the debugging needs.  In the usual circumstances (synth flow with no ICD features requested):

Reviewer 2: It is not clear to me that "module boundaries" is the proper place to introduce optimization barriers for debugging via a special cell. It is true that module interfaces are a common place where one may place debug probes (because the interfaces are usually well-defined, or at least more so than internal signals), but this is by no means an exclusive property, and usually you would also place debug probes on a subset of internal registers at least.

Also, it is not clear to me that *optimization barriers* are the right way to introduce debug probes even when you *do* want to observe signals at module boundaries. Optimization barriers are disruptive and inserting one on the critical path will degrade timing. If, instead, we used a CXXRTL-like method of reconstructing optimized out comb signals from registers only, the debug probes could be (entirely, or almost, depending on the architecture) free of impact on timing, which seems like a strictly superior technique.

Author: I don't intend to make module boundaries *special* for barrier purposes: named signals or registers work just as well for internal objects as for ports. I only go into detail about module boundaries here because they cause a bit of an annoying cartesian product: you can have a register that's also an output port, and you have to represent that as a single object (because otherwise you'd get a name collision).

As far as I can tell this is the exact same line of thought that resulted in Verilog having `output reg`.

- top-level ports:
  - top-level outputs: represented as a `signal` (with its output value unused); cannot be deleted
    - alternatively, a net with `keep` would effectively work the same
  - top-level inputs: represented as a `signal` (with input of `X`, as it is expected to be overriden); cannot be deleted

In a full-debug flow, the internal ports would be represented as signals (or registers) instead of nets.

### Named cells

A "named cell" associates a name with an instance, or a target cell.

```verilog
module top;
DSP48E2 mydsp([...]);
endmodule
```

```
!1 = scope "top"
!2 = cell "mydsp" scope=!1
%_ = target "DSP48E2" !2 {
    ...
}
```

A named cell can be marked as `keep`, which prevents the flow from deleting or transforming it.

## Type system

The final layer of debuginfo deals with describing the *types* of the named nets, signals, or registers.  The type system is only used for presentation, and has no effect on the synthesis flow — as far as synthesis semantics are concerned, the only type that exists is vectors of bits.

The following types are supported:

- simple types:
  - bool
  - unsigned integer

>   - signed integer

Reviewer 2: If we go as far as to have these two, we should support fixed point. DSP folks will thank us and it's not significantly more effort.

Author: I mean, it is a significant jump in the amount of effort needed to implement formatting, but yeah sure let's have that

  - abstract bitvector
- composite types:
  - enum
  - structure
  - tagged union
  - array

### Simple types

The `bool` type can be used only on single-bit values.  The values are displayed as `true` or `false`.

```
!1 = type bool
!2 = reg "a" type=!1
```

The `signed` and `unsigned` types can be used on values of any width.  The values are displayed as decimal by default.

The `bitvec` type can be used on values of any width.  The values are displayed as hex or binary.  If a named net or register has no metadata, the `bitvec` type is assumed.

### Enum types

> Matches the value to a predefined list to display it.  Has an associated width, and can only be used on values of that width.

Reviewer 2: Is it OK to have two names correspond to the same value, or should this be 1-to-1?

1-to-n mappings will come up in practice so if they are forbidden we should define or at least suggest a strategy for dealing with them in the frontends.

Author: I even included an ambigous name in the example!

yeah I don't think we can forbid them; let's just suggest the formatters pick the first one and be done with it

Reviewer 2: Oh I'm blind (I read it twice but missed it...)

### Array types

Used for memories, as well as packed arrays within structs.

```verilog
reg [31:0] gprs[32];
```

```
!1 = type bitvec
!2 = type array width=#32 depth=#32 type=!1
!3 = reg "gprs" width=#1024 type=!2
%0:_ = memory width=#32 depth=#32 {
    name !3:1024
    [...]
}
```

### Tagged union

Compares the value against a list of bit patterns, takes the first matching one, then decodes the value as a structure according to the matched variant.

```
!1 = type union #16 {
    011000xxxxxxxxxx = "ADD" {
        field "dst" start=#0 width=#5
        field "src" start=#5 width=#10
    }
    1011xxxxxxxxxxxx = "JUMP" {
        field "target" start=#0 width=#12
    }
    [...]
}
```

## Simulator interface

This RFC does not define a particular simulator or ICD interface.  However, we provide a guideline for the abstract operations such interfaces can provide.

A simulator interface has the following operations:

- hierarchically iterate available scopes and named objects
- get meta-information about scopes and named objects (names, types, widths, input/output marks, other attributes)
- read the current value of a named net
  - input: hierarchical name (must refer to a net)

>   - input: slice of the value to read (ie. can select all of the value, or just a subset of bits)

Reviewer 2: I think this part of the API might have to be redesigned depending on the outcome of https://github.com/prjunnamed/prjunnamed/pull/107/changes#r2869399223.

  - output: for each bit of the selected slice of the net, one of the following:
    - the value of the net (`0`, `1`, `X`)

>     - "value not available" error: there is no value in the netlist corresponding to the named net, or such value is inaccessible (eg. for ICD if the value is not at a location that can be tapped)

Reviewer 2: Should this not be per-bit? I can see only some bits being accessible in cases where e.g. people mix comb and sync domains for the same signal.

Author: I don't understand? the RFC does say this should be per bit

Reviewer 2: Oh I completely misread this (I think this is where I went and fixed the formatting) because it was kind of unreadable without scrolling).

- read the current value of a named signal: similar to the above

>   - TODO: if the signal is overriden, do we return the cell input, or the current override? both?

Reviewer 2: *If* we introduce the notion of named signals as described here, I think they should be treated as-if the bus was cut in this place and the halves routed to the toplevel (because this is exactly what you'd do when working with a system not supporting such a notion, or similarly in an actual ASIC).

  - input: value to be driven onto the signal slice (must have width matching signal)
  - from now on, the `signal` cell in the netlist will have the given value on output, until explicitly released or overriden again
- release a signal:
  - input: hierarchical name (must refer to a signal)
  - removes the override previously put on the signal, if any; from now on, the `signal` cell once again passes its input through to its output
- read the current value of a named register: similar to reading a named net or signal
- write the value of a named register
  - input: hierarchical name (must refer to a register)
  - input: slice of the register to write
  - input: value to put into the register slice
  - the register slice is instantaneously set to the given value, similar to what happens on an active and enabled clock edge
  - as opposed to overriding named signals, this does *not* result in a persistent override: the value written via the debug interface will be replaced by the next active clock edge, or similar triggers from the netlist
  - in particular, this is a no-op if an async reset is currently active (as the reset will immediately flip the changed value back)

>   - an error is returned if any of the selected register bits is not writable

Reviewer 2: These two things seem like they conflict with each other, seeing as an always-reset register can be transformed into a constant that isn't writable. Would this result in different behavior depending on the flavor of name and/or flow configuration?

- run target-specific or simulator-specific operation on a named cell
  - input: hierarchical name (must refer to a named cell)

>   - example: "set the analog input on this ADC cell to the given f64"

Reviewer 2: This is underspecified. I think there may be a need to do more complex scheduling, e.g. "perform this on the next posedge of `clk`", otherwise blackboxes may be quite difficult to interact with reliably.

I know this is just a guideline but this is something that causes serious problems in the current CXXRTL iteration so it may deserve some upfront attention.

- `cell <name or index> [scope=<scope>] [keep]`
- `type (bool | bitvec | unsigned | signed)`
- `type enum #<width> { <values> }`
- `type struct #<width> { <fields> }`
- `type union #<width> { <variants> }`
- `type array width=#<num> depth=#<num> [type=<type>]`

The following metadata items are removed:

- `ident` (replaced by `net` and `cell`)

The following cells are added to the IR:

- `%<id>:<width> = signal <value> <name or index> [scope=<scope>] [type=<type>] [lsb=#<num>] [from_msb] [input | output]`
- `%<id>:0 = name <value> <meta or meta slice>` (replaces current name cell)

The following cells are removed from the IR:

- `input` (replaced by `signal`)
- `output` (replaced by `signal`)
- `name` (replaced by the new `name` cell with different semantics)
- `debug` (replaced by `name`)

Furthermore, an optional `reg` metadata reference is added to the following cells:

- `dff`
- `memory` (memory contents and synchronous read registers)
- `target`

## Scopes

Scopes must form a tree.  A given hierarchical name must correspond to at most one scope or named object.

The IR puts no constraints on the names beyond being valid UTF-8 strings.

> Usually there will be exactly one top-level module scope containing all other scopes and named objects in the design (the `top` module).  However, this is not a hard requirement.  It is allowed to have multiple top-level scopes. It is also allowed, though heavily discouraged, to put named objects directly inside the root scope.

Reviewer 2: Is there a reason to allow this?

The following scope types are defined:

- `module` or `module(<name>)`: represents a Verilog `module` instance, a VHDL `entity` instance, or a similar construct; optionally, the name of the module type may be given for informational purposes
- `block`: represents a block within a module; corresponds to a Verilog `block`
- `struct` or `struct(<name>)`: represents a struct-typed signal, wire, or similar construct that got split into individual named nets or signals by the backend; the name of the struct type may be given for informational purposes
- `interface` or `interface(<name>)`: represents a grouping of module's input / output ports; like `module`, the name of the interface type may be given for informational purposes
- `array`: represents an array of blocks, interfaces, or module instances

> It is expected that indexed scopes will only be contained within a scope of type `array`, which may be a named scope or another (nested) indexed scope.  Likewise, it is expected that `array` scopes will contain only indexed scopes and named objects.  However, this is not a hard requirement.

Reviewer 2: Likewise, is there a reason to not make this a hard requirement? If we don't, each backend will have to deal with it individually, which seems bad.

The `interface` type does have hard validity requirements, I don't understand what requires `array` to be more relaxed.

- it can only contain (directly or indirectly): named nets, named signals, `interface` scopes, `array` scopes, `struct` scopes (cannot contain `block` or `module` scopes, nor other kinds of cells)

Netlist exporters are expected to deal with arbitrary names and scope hierarchies.  However, scope structures and names not representable in the target format or language may not be preserved (scopes may be flattened or inserted as necessary to legalize the hierarchy; names may be mangled, escaped, or stripped).  Therefore, unusual hierarchy structures and identifier characters should be avoided.

## Named nets

A named net is defined by the metadata item.  It has the following attributes:

- the name or index of the net (indexed nets are expected to be placed in `array` scopes)
- `scope=<scope>`: the parent scope, if any
- `width=#<int>`: the size of the wire, in bits (assumed to be 1 if absent)
- `lsb=#<int>`: the HDL-level index of the LSB, for display purposes (assumed to be 0 if absent); note that prjunnamed always internally counts bits from 0

> - `from_msb`: specifies that HDL-level bit indices grow from MSB up; therefore, physical bit `i` should be considered to have HDL-level index of `lsb - i` instead of `lsb + i`

Reviewer 2: These explanations would benefit from a clear definition of terms early in the RFC so that you don't need to refer to ambiguous "physical bit" and "prjunnamed internally".

  - `wire [0:63] a;` would be described as `lsb=63 from_msb`
- `src`: the source location of the net, as usual
- `type`: describes how to decode and display (or parse) values of the net

> - `keep`: disallows removing `name` cells referencing this net

Reviewer 2: I remarked on this before but to repeat: I think `keep` is very different from every other attribute and should probably be a part of the `name` cell itself.

- `input`: marks the net as an input of the closest containing `module` scope
- `output`: marks the net as an output of the closest containing `module` scope

The named net is associated with its value by means of `name` cells in the netlist.  A `name` cell has:

- no output
- input: a value of arbitrary width

> - associated named net or register

Reviewer 2: For registers, would you not sometimes end up with both `name` and `dff` cell referring to the metadata? This seems to be allowed. In this case, would a simulator not have difficulty going from metadata to cell, such as when it receives a read or write command?

Author: Yeah, we may end up with multiple `name` and `dff` cells for any given bit; this can and will come up from reasonable sequences of transformations. The simulator will need to handle that and maintain an index for that big many-to-many map.

- the slice (start + length) within that name or register

A single `name` cell associates a contiguous slice of a named net with a value.  For non-contiguous slices, multiple `name` cells can be used.

The mapping of named net bits to values is many-to-many:

- a given bit of named net may have no value in the netlist, representing a bit that has been deleted by the synthesis flow

> - a given bit of named net may have multiple values in the netlist, if the cells driving it have been duplicated by the flow

Reviewer 2: How should a simulator handle the case where two values for the same bit are different?

Author: I think that's a question of a given simulator's QoI and outside the scope of this RFC.

from semantics point of view, we *need* to represent multiple cells holding the same bit because we need to duplicate registers sometimes. while those are supposed to be equivalent, in practice they may end up holding different values at runtime (because the design invoked nondeterminism in some way). therefore, the situation is perfectly legal.

for a speed-optimized simulation, it'd be perfectly reasonable to just grab an arbitrary copy of the value. a more debug-oriented tool could notice the situation and show it as an `X` or some other special value

Reviewer 2: What I'm wondering here is whether we should adopt a representation that has a "canonical" value for a given name implied by the structure of the netlist, with other names attached to it as "aliases". This could be implicit in the numbering of metadata (but probably shouldn't), explicit in the netlist, explicit in metadata... a number of things. But I really don't feel comfortable about the "just grab a random, supposedly-equivalent register" plan because of the extremely disrputive way in which it will go wrong if there's a miscompilation.

Author: I don't think there's a way here to meaningfully improve the situation by having the netlist format bless a given copy instead of the simulator doing so. and "extremely disruptive way in which it will go wrong if there's a miscompilation"  is just... kind of how compilers work, isn't it?

Reviewer 2: Well, the benefit will be that all of the tooling agrees on which copy is blessed.

- a given netlist value may have multiple names, if the names were aliased in the netlist, or have been determined to be equivalent by the flow
- a given netlist value may have no names at all

## Named signals

> A named signal is defined by its cell.  Since it is a non-removable non-duplicatable optimization barrier, there is no need to split the definition into a metadata item.  It has the following attributes, which are essentially the same as for named nets:

Reviewer 2: It seems pretty painful for both the frontends and prjunnamed itself to have two incompatible ways of representing the same underlying information.

Author: yeahhh I think I'm just going to refactor signals to be more like a funny variant of registers

- name or index
- `scope=<scope>`
- `width=#<int>`
- `lsb=#<int>`
- `from_msb`
- `src`
- `type`
- `input`
- `output`

It is expected that design ports (which should be connected to I/O) are the ports of its top module.

## Named registers

A named register is defined in essentially the same way as a named net.  It has the following additional or modified attributes:

> - `keep_read`: disallows any transformation that would prevent externally reading the register (each bit of the register that has a mapping to the netlist must retain a mapping to the netlist)

Reviewer 2: What is "externally"? Are transformations that record how to reconstruct the register from other netlist entities, but outside of the netlist, allowed? (CXXRTL debug-eval-style)

Author: those are both things mentioned under future possibilities

Reviewer 2: I guess what I mean is that I'd like the RFC to pin this down one way or another.

- `keep_write`: disallows any transformation that would prevent externally writing the register (each bit of the register that is associated with a storage element must remain associated with a storage element unless it is determined to have no effect on the netlist and is removed entirely; ie. prevents adding bits to the `decayed` mask)
- `keep`: combined effect of `keep_read` and `keep_write`

> - `decayed=<mask>` marks a given subset of bits as "decayed", ie. non-writable and treated as named net instead of named register; such bits are associated with `name` cells in the netlist instead of storage elements

Reviewer 2: Why is this needed if it can be already determined by examining the netlist?

Is it not preferable to split the register to remove "decayed" bits or those represented by a `name` cell?

Author: splitting a register means we would no longer be able to apply a data format annotation to it; not having any good means of handling that is why the RFC allows mixed cells

the `decayed` mask cannot be accurately recovered from the netlist because of an edgecase: you may end up transforming a `dff` to a `name` (by determining it to be a constant or such), and then perform a transformation that makes assumption on its value, leaving no trace of the `name` in the netlist; in this case, a simulator needs to reject a write to that bit (even if there are still some `dff` bits referencing it), as it cannot be soundly performed. the purpose of the `decayed` mask is to record that event.
